mods/mc_tutorial_framework/init.lua

```diff
@@ -2,5 +2,150 @@ mc_tutorialFramework = { path = minetest.get_modpath("mc_tf")}
+
+		table.insert(names, #names + 1, name)
+		table.insert(descriptions, #descriptions + 1, description)
+
+		if schematic then
+			table.insert(schematics, #schematics + 1, schematic)
+		end
+end 
+
+addTutorial("Introduction", "Welcome to Minetest Classroom! To access tutorials, select the topic you would like to learn about on the left. Tutorials can also be accessed via portals that will teleport you to the tutorial relevant to the area you are in. To use a portal, stand in the wormhole until it transports you to a new area. Once you are in the tutorial realm, you can use the portal again to return to the area you were previously in.")
+addTutorial("Test", "testing", "testSchematic")
+addTutorial("Movement", "This tutorial explains how to walk in different directions, jump, and fly. To enter the tutorial, press the 'Teleport to Tutorial' button below. Once you are in the tutorial realm, you can use the portal again to return to the area you were previously in. If you need a reminder on how to use portals, go to 'Introduction'.", "movementTutorial")
+
+local function show_tutorial_menu(player)
+	if check_perm(player) then
+		local pname = player:get_player_name()
+
+		local formspec = ""
+		for i=1,#mc_tf_menu do 
+			formspec = formspec .. mc_tf_menu[i]
+		end
+
+		minetest.show_formspec(pname, "mc_tf:menu", formspec)
+		return true
+	end
+end
+
+-- The tutorial book for accessing tutorials
+minetest.register_tool("mc_tf:tutorialbook" , {
+	description = "Tutorial book",
+	inventory_image = "tutorial_book.png",
+	-- Left-click the tool activates the tutorial menu
+	on_use = function (itemstack, user, pointed_thing)
+        local pname = user:get_player_name()
+		-- Check for shout privileges
+		if check_perm(user) then
+			local textarea = mc_tf_menu[5]
+			textarea = "textarea[5,0.2;7.8,8;text;;" .. descriptions[1] .. "]"
+			mc_tf_menu[5] = textarea
+
+			show_tutorial_menu(user)
+		end
+	end,
+	-- Destroy the book on_drop to keep things tidy
+	on_drop = function (itemstack, dropper, pos)
+		minetest.set_node(pos, {name="air"})
+	end,
+})
+
+minetest.register_alias("tutorialbook", "mc_tf:tutorialbook")
+tutorialbook = minetest.registered_aliases[tutorialbook] or tutorialbook
+
+-- Processing the form from the menu
+minetest.register_on_player_receive_fields(function(player, formname, fields)
+	if string.sub(formname, 1, 5) ~= "mc_tf" then
+		return false
+	end
+
+	local wait = os.clock()
+	while os.clock() - wait < 0.05 do end --popups don't work without this
+
+	-- Menu
+	local event = minetest.explode_textlist_event(fields.tutorials)
+
+	if fields.tutorials then
+		if event.type == "CHG" then
+			local textarea = mc_tf_menu[5]
+			textarea = "textarea[5,0.2;7.8,8;text;;" .. descriptions[event.index] .. "]"
+			mc_tf_menu[5] = textarea
+			show_tutorial_menu(player)
+		end
+	end
+
+	if fields.teleport then
+		if event.index ~= 1 then
+			selectedRealm = mc_realmportals.CreateRealmByName(names[event.index + 1], schematics[event.index - 1])
```

Comment:
Can be efficient. Creating a new realm with schematics should be faster than generating new chunks using the terrain gen system. 

In terms of coordinate space property, creating a new realm for each tutorial instance is not efficient right now; will be decent soon. Being able to toggle whether it creates a new realm every time a player needs access or not would be good. The portal system currently works that way. Some tutorials (such as the node destruction one) needs to be reset every time it starts, whereas movement does not. 

I've been thinking about the best way to recycle coordinate space, and how we can recycle them efficiently. I'm trying to figure out something that takes minimal resources but is also flexible/resilient in case the server crashes.

For tutorials on campus, the following might work. We could use a table with selectedRealm in the book; On the table could be a function that is used when the player clicks on the "teleport" button. When we add new tutorials, we could supply a flag to define whether we teleport to a realm, coordinates,  open a formspec, etc.,

Reply:
> Some tutorials (such as the node destruction one) needs to be reset every time it starts, whereas movement does not.

I am thinking about the `Your Land` server (your-land.de:30000) that has a series of tutorials in the same building. The blocks are reset using a `mesecons_button`, probably triggering the placement of a schematic and then turning the player POV 180 degrees to stop the player from further interaction. This server does not use the `protector` mod, which is interesting. There is also a working example of giving players certain tools by hitting a `mesecons_button`, which could be a nice way to give players a clinometer, compass, etc.